src/librustc_lint/builtin.rs

```diff
@@ -1939,11 +1956,16 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidValue {
                                         "being left uninitialized"
                                     }
                                 ),
-                            )
-                            .note("this means that this code causes undefined behavior \
-                                when executed")
-                            .help("use `MaybeUninit` instead")
-                            .emit();
+                            );
+                            if let Some(span) = span {
+                                err.span_note(span, &msg);
+                            } else {
+                                err.note(&msg);
+                            }
+                            err.note("this means that this code causes undefined behavior \
+                                when executed");
+                            err.help("use `MaybeUninit` instead");
```

Comment:
I am not entirely sure about the order here. Should this note and help maybe be moved up above the thing that points to the struct/enum field?

Reply:
That's probably a good idea. Additionally, you could also try:
```rust
err.span_help(expr.span, "use `MaybeUninit<T>` instead");
```
to make the "suggestion" show inline (or `.span_label`) if that doesn't work.

Reply:
I don't think we have a good span to attach that message to, though.

Reply:
Hmm; why is `expr.span` not adequate here?

Reply:
I think only `span_suggestion` is shown inline, is it not? `span_note` certainly does not. Maybe `span_label` does.

My thinking was more to put the "References must be non-null" into the inline "suggestion".

Reply:
Okay I ave no idea what `span_label` does, but it completely messes things up here. The documentation isn't really explaining anything either. :/

Reply:
Yeah `span_label` can be shown inline (e.g. https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f0b319d7f7ad3d2d1c35014023c8c88a, the inline label there is from a `span_label`).

I think something like this would be nice:
```rust
error: the type `Wrap<&'static T>` does not permit zero-initialization
  --> $DIR/uninitialized-zeroed.rs:23:38
   |
LL |         let _val: Wrap<&'static T> = mem::zeroed();
   |                                      ^^^^^^^^^^^^^
   |                                      note: causes undefined behavior when executed
   |                                      help: use `MaybeUninit<T>` instead
   |
note: References must be non-null (in this struct field)
  --> $DIR/uninitialized-zeroed.rs:16:18
   |
LL | struct Wrap<T> { wrapped: T }
   |                  ^^^^^^^^^^
```

though I'm not exactly sure how to make the diagnostics infra show that. ^^

cc @estebank @oli-obk

Reply:
However... the beta cutoff is tomorrow so we could postpone further improvement for later and just do the reordering to put the note and the help above struct/enum field.

Reply:
Actually I was hoping for
```
error: the type `Wrap<&'static T>` does not permit zero-initialization
  --> $DIR/uninitialized-zeroed.rs:23:38
   |
LL |         let _val: Wrap<&'static T> = mem::zeroed();
   |                                      ^^^^^^^^^^^^^
   |                                      help: use `MaybeUninit<T>` instead
   |
note: causes undefined behavior when executed
note: References must be non-null
  --> $DIR/uninitialized-zeroed.rs:16:18
   |
LL | struct Wrap<T> { wrapped: T }
   |                  ^^^^^^^^^^
   |                  in this field
```
But maybe your placement of the "causes UB" is better.

Reply:
> However... the beta cutoff is tomorrow so we could postpone further improvement for later and just do the reordering to put the note and the help above struct/enum field.

Like this?

Reply:
Yeah the change is great; I'm happy for that to be the end state actually -- no follow up needed. :)